### PR TITLE
APP-88: [C] Integrate gpio demos for different platforms into single demo

### PR DESCRIPTION
**Pull request by RostakaGmfun**
/cc @Slash32 @vadimol

### CODE REVIEW

File: common/c/targets/common/target_gpio_led.h [hunk 1]
```diff
@@ -14,21 +14,29 @@
...
+/*
+ * Toggles output for GPIO LED id.
+ *
+ * Zero status mewans LOW, positive status means HIGH
+ */
+void target_gpio_led_toggle(int id, int status);
 
-/* Main task */
-void main_task(void *pvParameters);
+#endif // TARGET_SUPPORT_H
```
Comment (rasendubi) [inline common/c/targets/common/target_gpio_led.h]: comment is off

File: common/c/targets/common/target_gpio_led.h [hunk 2]
```diff
@@ -14,21 +14,29 @@
...
 
-/* WIFI */
-bool wifi_init();
-bool wifi_connect(const char *ssid, const char *pwd);
-void wifi_print_opmode();
-void wifi_print_station_config();
+/*
+ * Toggles output for GPIO LED id.
+ *
+ * Zero status mewans LOW, positive status means HIGH
```
Comment (rasendubi) [inline common/c/targets/common/target_gpio_led.h]: Zero -> 0
mewans -> means

File: common/c/targets/common/target_gpio_led.h [hunk 3]
```diff
@@ -14,21 +14,29 @@
...
+ *
+ */
 
-#include <stdbool.h>
+#ifndef TARGET_GPIO_LED_H
+#define TARGET_GPIO_LED_H
 
-/* UART */
-void uart_init();
+/*
```
Comment (rasendubi) [inline common/c/targets/common/target_gpio_led.h]: Why not doxygen

File: common/c/targets/esp8266/driver/gpio.h [hunk 4]
```diff
@@ -95,13 +95,13 @@ typedef struct {
 } GPIO_ConfigTypeDef;
 
 #define GPIO_OUTPUT_SET(gpio_no, bit_value) do { \
-    gpio_output_conf((bit_value)<<(gpio_no), ((~(bit_value))&0x01)<<(gpio_no), 1<<(gpio_no), 0) \
-    while(0);
+    gpio_output_conf((bit_value)<<(gpio_no), ((~(bit_value))&0x01)<<(gpio_no), 1<<(gpio_no), 0); \
+    } while(0);
```
Comment (rasendubi) [inline common/c/targets/esp8266/driver/gpio.h]: do-while(0) should be used as follows. (Note the absence of trailing semicolon.)

``` c
do {
    ...
} while (0)
```

Furthermore, do-while is actually not needed for single function invocation. The next would be enough:

``` c
#define GPIO_OUTPUT_SET(gpio_no, bit_value) \
    gpio_output_conf((bit_value)<<(gpio_no), ((~(bit_value))&0x01)<<(gpio_no), 1<<(gpio_no), 0)
```

Also, add spaces around operators.

File: common/c/targets/esp8266/driver/gpio.h [hunk 5]
```diff
@@ -95,13 +95,13 @@ typedef struct {
...
 
 #define GPIO_OUTPUT(gpio_bits, bit_value) do {\
     if(bit_value) gpio_output_conf((gpio_bits), 0, (gpio_bits), 0);\
     else gpio_output_conf(0, (gpio_bits), (gpio_bits), 0) \
-    while(0);
+    } while(0);
```
Comment (rasendubi) [inline common/c/targets/esp8266/driver/gpio.h]: Same here: remove semicolon

File: common/c/targets/esp8266/target.h [hunk 6]
```diff
@@ -42,6 +47,7 @@ int target_initialize(void);
  * but it is required to read user input in some demos,
  * so it is implemented here.
  */
+#undef getchar
```
Comment (rasendubi) [inline common/c/targets/esp8266/target.h:55]: Care to explain?

File: common/c/targets/esp8266/target_gpio_led.c [hunk 7]
```diff
@@ -14,24 +14,36 @@
...
+#define HIGH 1
+#define LOW 0
 
-/* Initialises a target. Zero value means success, negative - errors.
- *
- * For this particular target this will eventually try to connect to
- * the WiFi spot using SSID and password supplied during build.
- */
-int target_initialise(void);
+void target_gpio_led_init(void) {
```
Comment (rasendubi) [inline common/c/targets/esp8266/target_gpio_led.c]: Code style

File: common/c/targets/esp8266/target_gpio_led.c [hunk 8]
```diff
@@ -14,24 +14,36 @@
...
+    GPIO_OUTPUT_SET(0, LOW);
+    GPIO_OUTPUT_SET(2, LOW);
+}
 
-#endif //CC32XX_SUPPORT_H_
+void target_gpio_led_toggle(int id, int status) {
```
Comment (rasendubi) [inline common/c/targets/esp8266/target_gpio_led.c]: Code style

File: common/c/targets/esp8266/target_gpio_led.c [hunk 9]
```diff
@@ -14,24 +14,36 @@
...
+    if(id >= NUM_GPIO_LEDS || id < 0) {
```
Comment (rasendubi) [inline common/c/targets/esp8266/target_gpio_led.c]: code style

File: common/c/targets/esp8266/target_gpio_led.c [hunk 10]
```diff
@@ -14,24 +14,36 @@
...
+        return;
+    }
+    if(status) {
```
Comment (rasendubi) [inline common/c/targets/esp8266/target_gpio_led.c]: Code style

File: gpiocontrol/assembly/src-android.xml [hunk 11]
```diff
@@ -19,7 +19,7 @@
 <assembly xmlns="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2"
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:schemaLocation="http://maven.apache.org/plugins/maven-assembly-plugin/assembly/1.1.2 http://maven.apache.org/xsd/assembly-1.1.2.xsd">
-    <id>gpiocontrol-android-src</id>
+    <id>gpiocontrol-src-android</id>
```
Comment (rasendubi) [inline gpiocontrol/assembly/src-android.xml]: Not relevant to the PR

File: gpiocontrol/assembly/src-c.xml [hunk 12]
```diff
@@ -19,21 +19,21 @@
...
         <fileSet>
-            <directory>${project.basedir}/source/cc32xx</directory>
+            <directory>${project.basedir}/source/c</directory>
             <useDefaultExcludes>true</useDefaultExcludes>
-            <outputDirectory>gpio-cc-client</outputDirectory>
+            <outputDirectory>CGPIODemo</outputDirectory>
         </fileSet>
         <fileSet>
-            <directory>${main.dir}/common/nix/</directory>
+            <directory>${main.dir}/common/c/</directory>
```
Comment (rasendubi) [inline gpiocontrol/assembly/src-c.xml:39]: I wonder why nix is removed

File: gpiocontrol/source/c/CMakeLists.txt [hunk 13]
```diff
@@ -12,33 +12,40 @@
...
+    add_executable(${APP_NAME} src/kaa_demo.c)
+    target_link_libraries(${APP_NAME} kaac target_support)
+endif()
 
 # For most embedded targets binary stripping is required.
-if(${KAA_PRODUCE_BINARY})
-  add_custom_command(TARGET ${APP_NAME} POST_BUILD
-          COMMAND ${CMAKE_OBJCOPY} -O binary ${APP_NAME} ${APP_NAME}.bin)
-endif()
+if (${KAA_PRODUCE_BINARY})
```
Comment (ohamov) [inline gpiocontrol/source/c/CMakeLists.txt]: why did you add a space?

File: gpiocontrol/source/c/CMakeLists.txt [hunk 14]
```diff
@@ -12,33 +12,40 @@
...
+project(C-SDK-sample C)
 
 # Our application name.
-set(APP_NAME demo_client.afx)
+set(APP_NAME demo_client)
 
-# Demo executable.
-add_executable(${APP_NAME} src/kaa_demo.c)
+# Target-independent flags.
+set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu99 -g -Wall -Wextra")
```
Comment (rasendubi) [inline gpiocontrol/source/c/CMakeLists.txt]: We should stick with `-std=c99` and don't use gnu extensions as much as possible

File: gpiocontrol/source/c/CMakeLists.txt [hunk 15]
```diff
@@ -12,33 +12,40 @@
...
 # Directory containing target support library.
 add_subdirectory(targets/${KAA_TARGET})
+add_subdirectory(targets/common)
 
-# All libraries that demo app depends on.
-target_link_libraries(${APP_NAME} kaac target_support)
+# Demo executable.
+if("${KAA_TARGET}" STREQUAL "esp8266")
+    add_library(${APP_NAME}_s STATIC src/kaa_demo.c)
+    file(WRITE ${CMAKE_BINARY_DIR}/blank.c "")
```
Comment (rasendubi) [inline gpiocontrol/source/c/CMakeLists.txt]: Use `CMAKE_CURRENT_BINARY_DIR`

File: gpiocontrol/source/c/CMakeLists.txt [hunk 16]
```diff
@@ -12,33 +12,40 @@
...
 # Path to the Kaa SDK.
 add_subdirectory(libs/kaa)
 # Directory containing target support library.
 add_subdirectory(targets/${KAA_TARGET})
+add_subdirectory(targets/common)
 
-# All libraries that demo app depends on.
-target_link_libraries(${APP_NAME} kaac target_support)
+# Demo executable.
+if("${KAA_TARGET}" STREQUAL "esp8266")
```
Comment (rasendubi) [inline gpiocontrol/source/c/CMakeLists.txt]: Explanation is missing

File: gpiocontrol/source/c/CMakeLists.txt (hunk 13 above)
Comment (rasendubi) [inline gpiocontrol/source/c/CMakeLists.txt]: code style

File: gpiocontrol/source/c/build.sh [hunk 17]
```diff
@@ -43,12 +40,12 @@ KAA_C_LIB_HEADER_PATH="$KAA_LIB_PATH/src"
 KAA_CPP_LIB_HEADER_PATH="$KAA_LIB_PATH/kaa"
 KAA_SDK_TAR="kaa-c*.tar.gz"
 KAA_TOOLCHAIN_PATH_SDK=""
-# TODO APP-63: comment about inconvenience between KAA_TARGET and KAA_PLATFORM
-KAA_TARGET=cc32xx
-KAA_PRODUCE_BINARY=1
+# TODO: comment about inconvenience between KAA_TARGET and KAA_PLATFORM
```
Comment (rasendubi) [inline gpiocontrol/source/c/build.sh]: Should be resolved

File: gpiocontrol/source/c/build.sh [hunk 18]
```diff
@@ -57,8 +54,32 @@ echo " ACCESS TOKEN: " $DEMO_ACCESS_TOKEN
 echo "==================================="
 
 
+select_arch() {
+    echo "Please enter a target:"
+    read target
+
+    # TODO: better case handling
```
Comment (rasendubi) [inline gpiocontrol/source/c/build.sh]: Resolve this TODO item or open an issue.

File: common/c/targets/common/target_gpio_led.h [hunk 19]
```diff
@@ -14,21 +14,29 @@
  *  limitations under the License.
  */
 
-#ifndef ESP_PLATFORM_H
-#define ESP_PLATFORM_H
+/*
+ * This header provides several bindings for TI's cc32xx target that abstracts
+ * an implementation of its features. Right now it contains only
+ * console and target initialisation routines, but it must be extended
```
Comment (ohamov) [inline common/c/targets/common/target_gpio_led.h]: `s/initialisation/initialization`

File: gpiocontrol/source/c/build.sh [hunk 20]
```diff
@@ -57,8 +54,32 @@ echo " ACCESS TOKEN: " $DEMO_ACCESS_TOKEN
...
+        KAA_TARGET=$target
+        KAA_PRODUCE_BINARY=true
+        KAA_REQUIRE_CREDENTIALS=true
+        ;;
+    esac
+}
+
 unpack_sdk() {
-    if [ ! -d "$KAA_C_LIB_HEADER_PATH"  -a ! -d "$KAA_CPP_LIB_HEADER_PATH" ]
+    if [[ ! -d "$KAA_C_LIB_HEADER_PATH" &&  ! -d "$KAA_CPP_LIB_HEADER_PATH" ]]
```
Comment (rasendubi) [inline gpiocontrol/source/c/build.sh]: AFAIK, there is no `[[` in plain sh.

File: gpiocontrol/source/c/build.sh [hunk 21]
```diff
@@ -74,25 +95,20 @@ unpack_sdk() {
...
     mkdir -p "$PROJECT_HOME/$BUILD_DIR"
     cd $BUILD_DIR
 
-    echo "Enter WiFi SSID:"
-    read SSID
-    echo "Enter WiFi Password:"
-    read PASSWORD
-
-    KAA_TOOLCHAIN_PATH_SDK="-DCMAKE_TOOLCHAIN_FILE=$KAA_LIB_PATH/toolchains/cc32xx.cmake"
+    if [[ $KAA_REQUIRE_CREDENTIALS = true ]]
```
Comment (rasendubi) [inline gpiocontrol/source/c/build.sh]: See comment above on `[[`.

File: common/c/targets/common/target_gpio_led.h (hunk 19 above)
Comment (ohamov) [reply common/c/targets/common/target_gpio_led.h]: `s/must/can` or remove `if required`

File: gpiocontrol/source/c/build.sh [hunk 22]
```diff
@@ -114,11 +132,16 @@ clean() {
 }
 
 run() {
-    echo "To run demo, please have a look at http://docs.kaaproject.org/display/KAA/Texas+Instruments+CC3200#TexasInstrumentsCC3200-Example"
+    cd "$PROJECT_HOME/$BUILD_DIR"
+    ./$APP_NAME
 }
 
-case "$1" in
+for cmd in $@
```
Comment (rasendubi) [inline gpiocontrol/source/c/build.sh]: Use next

``` sh
for cmd in "$@"; do
```

File: gpiocontrol/source/c/build.sh [hunk 23]
```diff
@@ -114,11 +132,16 @@ clean() {
...
+do
+
+case "$cmd" in
```
Comment (rasendubi) [inline gpiocontrol/source/c/build.sh]: Add indentation level.

File: gpiocontrol/source/c/src/kaa_demo.c [hunk 24]
```diff
@@ -52,12 +47,12 @@ static void kaa_device_info_request(void *context
 
     kaa_remote_control_ecf_device_info_response_t *response = kaa_remote_control_ecf_device_info_response_create();
 
-    response->device_name = kaa_string_copy_create("CC3200");
-    response->model       = kaa_string_copy_create("LaunchPad");
+    response->device_name = kaa_string_copy_create(TARGET_DEVICE_NAME);
+    response->model       = kaa_string_copy_create(TARGET_MODEL_NAME);
     response->gpio_status = kaa_list_create();
 
     int i = 0;
```
Comment (rasendubi) [inline gpiocontrol/source/c/src/kaa_demo.c]: Inline `i`

File: gpiocontrol/source/c/src/kaa_demo.c [hunk 25]
```diff
@@ -79,24 +74,25 @@ static void kaa_GPIOToggle_info_request(void *context
...
-    if (event->gpio->status) {
-        GPIO_IF_LedOn(MCU_RED_LED_GPIO + event->gpio->id);
-        gpio_led[event->gpio->id] = 1;
-    } else {
-        GPIO_IF_LedOff(MCU_RED_LED_GPIO + event->gpio->id);
-        gpio_led[event->gpio->id] = 0;
-    }
+    gpio_led[event->gpio->id] = event->gpio->status;
+
+    int gpio_id = event->gpio->id >= NUM_GPIO_LEDS?NUM_GPIO_LEDS:event->gpio->id;
```
Comment (rasendubi) [inline gpiocontrol/source/c/src/kaa_demo.c]: Add spaces around operators

Comment (rasendubi) [reply gpiocontrol/source/c/src/kaa_demo.c]: Also, add braces around condition (according to MISRA)

**Review by Slash32 — CHANGES_REQUESTED** [review]
The comments must be addressed before merge.

**Comment by kaa-leeroy** [comment]
Can one of the admins verify this patch?

File: gpiocontrol/source/c/build.sh [hunk 26]
```diff
@@ -57,53 +67,55 @@ echo " ACCESS TOKEN: " $DEMO_ACCESS_TOKEN
...
     mkdir -p "$PROJECT_HOME/$BUILD_DIR"
     cd $BUILD_DIR
 
-    echo "Enter WiFi SSID:"
-    read SSID
-    echo "Enter WiFi Password:"
-    read PASSWORD
-
-    KAA_TOOLCHAIN_PATH_SDK="-DCMAKE_TOOLCHAIN_FILE=$KAA_LIB_PATH/toolchains/cc32xx.cmake"
+    if [ $KAA_REQUIRE_CREDENTIALS = true ]
```
Comment (rasendubi) [inline gpiocontrol/source/c/build.sh]: Code style:

``` sh
if [ $KAA_REQUIRE_CREDENTIALS = true ]; then
```

File: common/c/targets/cc32xx/target_gpio_led.c [hunk 27]
```diff
@@ -0,0 +1,52 @@
+/**
```
Comment (rasendubi) [inline common/c/targets/cc32xx/target_gpio_led.c]: Should not be doxygen comment

**Review by rasendubi — COMMENTED** [review]
Minor. Otherwise, looks good.

Have you tested all platforms after changing std to c99? This could break something.

**Comment by RostakaGmfun** [comment]
@rasendubi demo compiles fine for all targets with `c99`.

File: common/c/targets/cc32xx/target.h [hunk 28]
```diff
@@ -14,10 +14,10 @@
  *  limitations under the License.
  */
 
-/*
+/**
```
Comment (rasendubi) [inline common/c/targets/cc32xx/target.h:17]: That's file-level doxygen comment: you must add `@file`.

File: common/c/targets/common/target_gpio_led.h [hunk 29]
```diff
@@ -14,29 +14,31 @@
...
  */
 
-/*
- * This header provides several bindings for TI's cc32xx target that abstracts
- * an implementation of its features. Right now it contains only
- * console and target initialisation routines, but it must be extended
- * if required.
- *
+/**
+ * This header defines generic API to access
```
Comment (rasendubi) [inline common/c/targets/common/target_gpio_led.h:20]: That's file-level doxygen comment: you must add `@file`.

File: common/c/targets/esp8266/target.h [hunk 30]
```diff
@@ -14,10 +14,10 @@
  *  limitations under the License.
  */
 
-/*
+/**
```
Comment (rasendubi) [inline common/c/targets/esp8266/target.h:17]: That's file-level doxygen comment: you must add `@file`.

File: gpiocontrol/source/c/src/kaa_demo.c [hunk 31]
```diff
@@ -76,7 +75,7 @@ static void kaa_GPIOToggle_info_request(void *context
 
     gpio_led[event->gpio->id] = event->gpio->status;
 
-    int gpio_id = event->gpio->id >= NUM_GPIO_LEDS?NUM_GPIO_LEDS:event->gpio->id;
+    int gpio_id = event->gpio->id >= (NUM_GPIO_LEDS ? NUM_GPIO_LEDS : event->gpio->id);
```
Comment (rasendubi) [inline gpiocontrol/source/c/src/kaa_demo.c]: This changes semantics. The previous expression was parsed as:

``` c
int gpio_id = (event->gpio->id >= NUM_GPIO_LEDS) ? NUM_GPIO_LEDS : event->gpio->id;
```

That's what I meant by "around condition".

Comment (rasendubi) [reply gpiocontrol/source/c/src/kaa_demo.c]: Looked into meaning. Could be replaced by `min` function/macro.

**Comment by Slash32** [comment]
@kaa-leeroy test this hard please

**Review by rasendubi — APPROVED** [review]

**Review by Slash32 — APPROVED** [review]

**Comment by Slash32** [comment]
@kaa-leeroy test this hard please

File: common/c/targets/esp8266/target.h (hunk 6 above)
Comment (rasendubi) [reply common/c/targets/esp8266/target.h:55]: How this worked before?

**Review by Slash32 — APPROVED** [review]

**Pull request merged**